### PR TITLE
VSA support for Wan 2.2 and LTX2

### PR DESCRIPTION
**Pull request by jingyu-ml**
### What does this PR do?

Type of change: new feature

Adds end-to-end **Video Sparse Attention (VSA)** inference support for **Wan 2.2**
(5B and 14B) under `modelopt.torch.sparsity.attention_sparsity` (`mtsa`). The
core VSA method landed in #1053 but the original LTX-2 plugin was dropped and
Wan 2.2 had none, so neither model was actually runnable with VSA. This PR
fills that gap and unifies the Wan 2.2 example around a single
`--method {skip_softmax,vsa}` entry point.

Main changes:

- **New plugin `plugins/wan22.py`** — forward pre-hook on `WanTransformer3DModel`
  that reads `hidden_states.shape = (B, C, T, H, W)`, divides by
  `config.patch_size`, and propagates the post-patchify `(T, H, W)` to every
  `SparseAttentionModule` via `method.set_video_shape()`. Wan uses
  `F.scaled_dot_product_attention`, so VSA's existing SDPA patch handles the
  rest — no module subclass needed.
- **VSA `gate_compress=None` fix** (`methods/vsa.py`) — the fastvideo kernel's
  default `compress_attn_weight=None` returns `out_c + out_s`, which doubles
  the attention signal on any model without a learned gate (e.g. Wan 2.2). VSA
  now passes an explicit `gate=0` tensor so `out = 0 * out_c + out_s = out_s`.
  Side effect: `top_k_ratio=1.0` now cleanly degenerates to dense SDPA
  (modulo bf16 rounding).
- **Plugin registry** (`plugins/__init__.py`) — `CUSTOM_MODEL_PLUGINS` changed
  from `list` to `set` so re-imports stay idempotent (matches quantization /
  peft convention). Wan 2.2 plugin registered via `import_plugin` so a missing
  optional dep never breaks the core sparse-attention API.
- **Example unification** (`wan22_skip_softmax.py` → `wan22_sparse_attn.py`) —
  single script with `--method {skip_softmax,vsa}` plus VSA flags
  (`--top-k-ratio`, `--skip-first-last`, `--enable-vae-tiling`). Skip-softmax
  behaviour and CLI are preserved.
- **README rewrite** (`examples/diffusers/sparsity/README.md`) — method
  comparison table, VSA quick-start, config reference, and a dense-equivalence
  sanity-check section with measured PSNR numbers on Wan 2.2 14B.

LTX-2 plugin (`plugins/ltx2.py`) is included as well — it wraps LTX-2's native
`LTXSelfAttention` and calls `VSA.forward_attention` directly, with a
zero-initialised trainable `to_gate_compress` — but the LTX-2 example is
**not** in this PR (it depends on third-party `ltx_core` / `ltx_trainer` /
`ltx_pipelines` under the LTX Community License). Example will land separately
once the training loop and license plumbing are finalised.

### Usage

```python
import torch
from diffusers import AutoencoderKLWan, WanPipeline
import modelopt.torch.sparsity.attention_sparsity as mtsa

pipe = WanPipeline.from_pretrained(
    "Wan-AI/Wan2.2-T2V-A14B-Diffusers", torch_dtype=torch.bfloat16
).to("cuda")

# VSA config: 50% top-K on self-attention, cross-attention left dense.
# ``video_shape`` is auto-derived by the Wan 2.2 plugin on each forward.
config = {
    "sparse_cfg": {
        "*.attn1*": {
            "method": "vsa",
            "block_size_3d": (4, 4, 4),
            "top_k_ratio": 0.5,
            "enable": True,
        },
        "*.attn2*": {"enable": False},
        "default": {"enable": False},
    },
}
pipe.transformer = mtsa.sparsify(pipe.transformer, config)

video = pipe(prompt="A cat playing piano", num_frames=81).frames[0]
```

Or the built-in default via the example script:

```bash
# VSA at 50% top-K (default block_size_3d=(4,4,4), self-attn only)
python examples/diffusers/sparsity/wan22_sparse_attn.py --method vsa \
    --top-k-ratio 0.5 \
    --prompt "A cat playing piano" --output vsa.mp4

# Skip-softmax (unchanged behaviour, still the default method)
python examples/diffusers/sparsity/wan22_sparse_attn.py \
    --raw-threshold -0.7 \
    --prompt "A cat playing piano" --output out.mp4
```

### Testing

- **Unit tests** — `conda run -n modelopt python -m pytest
  tests/unit/torch/sparsity/attention_sparsity/` → 149 passed (sparse-attention
  conversion, kernel backends, registry).
- **Wan 2.2 plugin hook test** — end-to-end check that `video_shape` is
  correctly derived from `hidden_states.shape / patch_size` and propagated
  to every VSA method instance before the SDPA patch fires.
- **Dense-equivalence sanity check on Wan 2.2 14B** (720×1280 / 81 frames
  / 40 steps), first-frame PSNR vs dense baseline:

  | Comparison | PSNR |
  |---|---|
  | baseline vs baseline w/ VAE tiling | 40.5 dB |
  | baseline vs VSA `top_k_ratio=1.0`  | 23.9 dB |
  | baseline vs VSA `top_k_ratio=0.5`  | 13.1 dB |

  The ~24 dB drop at `top_k_ratio=1.0` is error accumulation over 6400
  attention calls through the denoising loop; single-call PSNR vs dense
  SDPA on random inputs is ~50 dB, confirming the dense-equivalence property
  at the kernel level.
- **No regression on skip-softmax** — existing Wan 2.2 skip-softmax flows
  (raw threshold, calibration, dense Triton baseline) verified through the
  renamed `wan22_sparse_attn.py` script.

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ 
- Did you write any new necessary tests?: ✅ 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ 

## Summary by CodeRabbit

* **New Features**
  * Added VSA sparse-attention alongside Skip‑Softmax with CLI method switching, VSA options (block size, top-k ratio, video-shape) and an option to enable VAE tiling.
  * Automatic VSA support for Wan‑2.2 and LTX‑2 models with inferred video-shape propagation and safer handling when shapes are unavailable.
  * Improved runtime sparsity reporting behavior per method.

* **Documentation**
  * Rewrote sparsity README with quick-starts, method comparisons, model-to-script mapping, defaults and known issues.

* **Tests**
  * Added unit tests for LTX‑2 and Wan‑2.2 VSA plugins and video-shape extraction/behavior.

* **Chores**
  * Plugin registration deduplicated to prevent duplicate callbacks; minor robustness fixes in VSA handling.

### CODE REVIEW

File: modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (cjluo-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py:1]: Copyright year should be 2025 for new files (per repo convention).

File: modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (cjluo-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py:1]: Copyright year should be 2025 for new files (per repo convention).

File: modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        def _noop(module, args, kwargs):
+            return None
+
+        return _noop
+
+    p_t, p_h, p_w = patch_size
+
+    def _hook(module: nn.Module, args: tuple, kwargs: dict) -> None:
+        hidden_states = _extract_hidden_states(args, kwargs)
+        if hidden_states is None or hidden_states.ndim != 5:
```
Comment (cjluo-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py:117]: Minor: `_noop` is registered with `with_kwargs=True` but returns `None`. For pre-hooks with `with_kwargs=True`, returning `None` is fine (no-op), but it's worth adding a brief comment noting that `None` return means "don't modify args/kwargs".

File: modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+        value = self._reshape_for_vsa(value, self.heads)
+        if gate_compress is not None:
+            gate_compress = self._reshape_for_vsa(gate_compress, self.heads)
+
+        output, stats = method.forward_attention(
+            query=query,
+            key=key,
+            value=value,
+            gate_compress=gate_compress,
+            video_shape=video_shape,
```
Comment (cjluo-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py:310]: The `_call_original_forward` pattern of temporarily toggling `self._enabled` is functional but fragile — if `SparseAttentionModule.forward` changes to check `is_enabled` via a different path, this breaks silently. Consider adding a comment noting this coupling, or checking whether `super().forward()` (from `DynamicModule`) could be called directly instead.

File: modelopt/torch/sparsity/attention_sparsity/plugins/__init__.py [hunk 5]
```diff
@@ -15,19 +15,32 @@
...
         callback(model)
 
 
+# Built-in plugins
 from . import huggingface  # noqa: E402
 
+# Model-specific plugins for VSA.  Guarded by ``import_plugin`` so the
+# module-level imports stay soft — a missing dependency in one plugin must
+# not break the core sparse-attention API.
+with import_plugin("ltx2"):
```
Comment (cjluo-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/__init__.py:42]: The `import_plugin` guards here are defensive but note that neither `wan22.py` nor `ltx2.py` import any optional third-party packages at module level — they only use `torch.nn` and internal modelopt imports. The `ltx_core` import in ltx2 is inside a function. So these guards will never actually catch a `ModuleNotFoundError`. This is fine for future-proofing, but worth noting in the comment so future readers don't wonder what dependency is being guarded.

**Review by cjluo-nv — COMMENTED** [review]
This PR adds VSA support for Wan 2.2 and LTX-2 models. The core logic looks correct and well-documented. The VSA `gate_compress=None` fix is important and well-reasoned. The plugin architecture follows good patterns (idempotent registration, graceful fallbacks, class-name-based detection). 

Key observations:
1. **Copyright year**: Both new files (`ltx2.py`, `wan22.py`) use "Copyright (c) 2024" but should be 2025 for new files.
2. **Missing plugin-specific tests**: While `test_vsa.py` covers the core VSA method, config validation, and integration, there are no unit tests for the wan22 or ltx2 plugins specifically (hook installation, video_shape extraction, idempotency guards). The PR description mentions a "Wan 2.2 plugin hook test" but it doesn't appear in the test files.
3. **The LTX-2 plugin is included without an example**: This is explicitly called out in the PR description as intentional (pending license/training loop), which is fine, but makes the lack of tests more concerning since there's no way to validate it even manually.
4. **Size**: ~1052 lines is at the boundary but the changes are cohesive.

The code quality is high overall — good docstrings, defensive fallbacks, and proper error messages. The `CUSTOM_MODEL_PLUGINS` list→set migration is a nice improvement.

**Comment by cjluo-nv** [comment]
@kaix-nv could you take a look?

**Review by kevalmorabia97 — APPROVED** [review]
LGTM from LTX license notice point of view

File: modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py (hunk 1 above)
Comment (jingyu-ml) [reply modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py:1]: should be 2024, just like other files

File: modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py (hunk 2 above)
Comment (jingyu-ml) [reply modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py:1]: should be 2024, just like other files

**Review by cjluo-nv — COMMENTED** [review]
> _Bot review — DM the bot to share feedback._

**Re-review summary**: All critical issues from the previous review have been addressed:

1. ✅ **calibrate.py prog_idx flattening** — Fixed; now uses `tl.num_programs(2)` instead of `tl.load(b_seq_len + 0)`.
2. ✅ **vsa.py gate_compress=None docstring** — Updated to document the new "disabled/zero" semantics.
3. ✅ **wan22.py hook unconditionally overwriting video_shape** — Now uses `_wan22_auto_video_shape` marker to preserve user-supplied shapes.
4. ✅ **Missing plugin-specific tests** — Comprehensive unit tests added for both plugins (`test_wan22_plugin.py`, `test_ltx2_plugin.py`) covering detection, hook registration, idempotency, shape extraction, and edge cases.

**Design**: The PR extends the existing `attention_sparsity` plugin architecture with two new model-specific plugins (Wan 2.2, LTX-2) and a new VSA method. No new subsystem or framework is introduced — it plugs into the existing `SparseAttentionMethod` registry and `CUSTOM_MODEL_PLUGINS` set. This is the right approach.

**Remaining minor items** (not blocking, but worth owner attention):
- The `huggingface` plugin is still eagerly imported in `plugins/__init__.py` (not guarded by `import_plugin`), while `ltx2` and `wan22` are. This inconsistency was flagged in the prior review and is still present. Low risk since HF is effectively a hard dep, but worth aligning.
- The threshold validation in `calibrate.py` (rejecting `t <= 0` before `math.log2`) was suggested but not added. Minor defensive improvement.
- Size is ~1500 lines but cohesive: two plugins, one method fix, one example refactor, two test files, and a README rewrite. Reasonable not to split.

Nudging for human sign-off since this is a large, multi-file change touching kernel code and adding new plugin integrations.

File: examples/diffusers/sparsity/README.md [hunk 6]
```diff
@@ -26,51 +47,202 @@ Two modes are supported:
...
+1. **Compression branch** — averages tokens within a 3D block (default `4,4,4` =
+   64 tokens) and computes coarse-grained block-level attention for global context.
+2. **Sparse branch** — selects the top-K most important blocks by the compression
+   branch's attention scores and computes fine-grained attention only on those.
+3. **Gate blend** — `output = compression * gate_compress + sparse`. On models
+   without a learned `gate_compress` (Wan 2.2, and LTX-2 until fine-tuned), VSA
+   passes a zero tensor so `output = 0 * compression + sparse = sparse`. This
+   makes VSA at `top_k_ratio=1.0` (keep all blocks) mathematically equivalent to
+   dense attention, modulo `bfloat16` kernel rounding (~10⁻⁵ per call on a 75k
+   token sequence).
```
Comment (kaix-nv) [inline examples/diffusers/sparsity/README.md:114]: remove?

File: examples/diffusers/sparsity/README.md [hunk 7]
```diff
@@ -26,51 +47,202 @@ Two modes are supported:
...
+
+VSA is **calibration-free** — sparsity is controlled by a fixed `top_k_ratio`
```
Comment (kaix-nv) [inline examples/diffusers/sparsity/README.md:116]: Does skip softmax also use `top_k_ratio` for specifying sparsity ratio?

File: examples/diffusers/sparsity/README.md [hunk 8]
```diff
@@ -13,11 +13,32 @@
...
+| **Skip-Softmax** (BLASST) | Drop low-impact KV tiles inside FlashAttention. Works on any transformer with bidirectional attention. | Optional (exponential model) |
+| **VSA** (Video Sparse Attention) | Block-level two-branch attention tuned for video models with long 3D token sequences. | None (fixed `top_k_ratio`) |
+
+Switching between methods is a CLI/config change — the pipelines, APIs,
+and plugins are shared.
+
+| Model | Script | Methods |
+|-------|--------|---------|
+| Wan 2.2 5B / 14B | `wan22_sparse_attn.py` | `--method skip_softmax` (default), `--method vsa` |
+| LTX-2            | `ltx2_vsa.py`          | VSA only (LTX-2 uses a custom attention module; skip-softmax backend in progress) |
```
Comment (kaix-nv) [inline examples/diffusers/sparsity/README.md:30]: Do we still have `ltx2_vsa.py`?

File: examples/diffusers/sparsity/README.md [hunk 9]
```diff
@@ -26,51 +47,202 @@ Two modes are supported:
...
+    --text-encoder-path /path/to/gemma \
+    --top-k-ratio 0.5 \
+    --prompt "A cat playing piano" --output vsa.mp4
+
+# LTX-2 — baseline (no VSA)
+python ltx2_vsa.py \
+    --checkpoint /path/to/ltx2.safetensors \
+    --text-encoder-path /path/to/gemma \
+    --no-vsa --output baseline.mp4
+```
```
Comment (kaix-nv) [inline examples/diffusers/sparsity/README.md:154]: Same. Is `ltx2_vsa.py` renamed?

File: modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+        # Incompatible seq_len (e.g., audio attention with seq=32)
+        seq_len = query.shape[1]
+        block_size_3d = method.block_size_3d
+        block_elements = block_size_3d[0] * block_size_3d[1] * block_size_3d[2]
+        if seq_len < block_elements:
+            logger.debug(f"VSA skipped: seq_len={seq_len} < block_elements={block_elements}")
+            return self._call_original_forward(*args, **kwargs)
+
+        video_shape = self._resolve_video_shape(seq_len)
```
Comment (kaix-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py:289]: If `video_shape` is `None`, what behavior do we expect? Should we raise an error instead, or let the user run in dense mode?

File: modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+
+        # Weakref avoids the circular-submodule problem (nn.Module.__setattr__
+        # would otherwise register the root model as a submodule of every
+        # attention, causing infinite recursion in named_children()).
+        root_ref = weakref.ref(model)
+        for _, module in model.named_modules():
+            if _is_ltx2_attention_module(module):
+                object.__setattr__(module, "_vsa_root_model_ref", root_ref)
+
+        model.register_forward_pre_hook(_extract_video_shape_hook)
```
Comment (kaix-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/ltx2.py:395]: We need a sentinel to avoid calling `register_forward_pre_hook` multiple times. This exists in WAN but not in LTX2.
```
for transformer in transformers:
      if getattr(transformer, "_vsa_hook_registered", False):
          continue
      hook = _make_wan22_video_shape_hook(transformer)
      transformer.register_forward_pre_hook(hook, with_kwargs=True)
```

File: modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+2. Divides by ``model.config.patch_size = (p_t, p_h, p_w)`` — same
+   computation diffusers does internally (see
+   ``WanTransformer3DModel.forward``: ``post_patch_num_frames = num_frames // p_t``
+   etc.).
+3. Propagates the resulting shape to every ``SparseAttentionModule`` in
+   the transformer whose method is VSA, via ``method.set_video_shape()``.
+
+Self-attention layers (``attn1``) then see a valid ``video_shape`` when the
+SDPA patch fires.  Cross-attention (``attn2``) is skipped by VSA's
+``can_apply_vsa`` guard since Q/K lengths differ.
```
Comment (kaix-nv) [inline modelopt/torch/sparsity/attention_sparsity/plugins/wan22.py:38]: Is this True for WAN?

File: tests/unit/torch/sparsity/attention_sparsity/test_ltx2_plugin.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (kaix-nv) [inline tests/unit/torch/sparsity/attention_sparsity/test_ltx2_plugin.py:1]: Should we add save/restore test for vsa?

**Comment by kaix-nv** [comment]
@jingyu-ml Overall LGTM. Left some comments.

**Review by shengliangxu — APPROVED** [review]
LGTM
